lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart

```diff
@@ -603,128 +578,108 @@ class HtmlViewEmbedder {
       // to the backup surface.
       SurfaceFactory.instance.releaseSurfaces();
       _overlays.clear();
-      final List<int> viewsNeedingOverlays = _compositionOrder
-          .where((int viewId) => platformViewManager.isVisible(viewId))
-          .toList();
-      final int numOverlays = math.min(
-        SurfaceFactory.instance.maximumOverlays,
-        viewsNeedingOverlays.length,
-      );
-      for (int i = 0; i < numOverlays; i++) {
-        final int viewId = viewsNeedingOverlays[i];
-        assert(!_viewsUsingBackupSurface.contains(viewId));
-        _initializeOverlay(viewId);
-      }
-      _assertOverlaysInitialized();
-      return null;
+      viewsNeedingOverlays.forEach(_initializeOverlay);
     } else {
       // We want to preserve the overlays in the "unchanged" section of the
-      // diff result as much as possible. If `addToBeginning` is `false`, then
-      // release the overlays for the views which were deleted from the
-      // beginning of the composition order and try to reuse those overlays in
-      // either the unchanged segment or the newly added views. If
-      // `addToBeginning` is `true` then release the overlays for the deleted
-      // views and from the unchanged segment and assign the newly added views
-      // to them.
-      diffResult.viewsToRemove.forEach(_releaseOverlay);
-      final int availableOverlays =
-          SurfaceFactory.instance.numAvailableOverlays;
-      if (diffResult.addToBeginning) {
-        // If we have enough overlays for the newly added views, then just use
-        // them. Otherwise, we will need to release overlays from the unchanged
-        // segment of view ids.
-        if (diffResult.viewsToAdd.length > availableOverlays) {
-          int viewsToDispose = math.min(SurfaceFactory.instance.maximumOverlays,
-              diffResult.viewsToAdd.length - availableOverlays);
-          // The first `maximumSurfaces` views in the previous composition order
-          // had an overlay.
-          int index = SurfaceFactory.instance.maximumOverlays -
-              diffResult.viewsToAdd.length;
-          while (viewsToDispose > 0) {
-            // The first [maxOverlays - viewsAdded] active views should have
-            // overlays. The rest should be removed.
-            _releaseOverlay(_activeCompositionOrder[index++]);
-            viewsToDispose--;
-          }
-        }
-
-        // Now assign an overlay to the newly added views.
-        final int overlaysToAssign = math.min(diffResult.viewsToAdd.length,
-            SurfaceFactory.instance.numAvailableOverlays);
-        for (int i = 0; i < overlaysToAssign; i++) {
-          _initializeOverlay(diffResult.viewsToAdd[i]);
+      // diff result as much as possible. Iterate over all the views needing
+      // overlays and assign them an overlay if they don't have one already.
+      final List<int> viewsWithOverlays = _overlays.keys.toList();
```

Comment:
I can't because the for-loop modifies `_overlays`. It will throw an exception if you modify a map while iterating over its keys, so we have to basically take a "snapshot" of the keys and iterate over those.